[Lite Interpreter] Enable __setstate__

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33294 [Lite Interpreter] Enable __setstate__**

1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)

Commit:
register ops for quantized linear.

Commit:
Update on "register ops for quantized linear."

Commit:
Update on "register ops for quantized linear."

Commit:
Update on "register ops for quantized linear."

kimishpatel commented on torch/csrc/jit/export_module.cpp:
```diff
@@ -24,7 +24,86 @@ namespace {
+  }
+
+  // operators
+  std::vector<IValue> operators;
```
Given we know the vector size, we can just reserve that here?

kimishpatel commented on torch/csrc/jit/export_module.cpp:
```diff
@@ -24,7 +24,86 @@ namespace {
-};
+}
+
+static IValue Tup(std::vector<IValue> ivalues) {
```
Would it make sense to have a version of this that accepts rvalue ref? I see call-sites for this where passed in vector is not used afterwards and we still end up copying it during the call to this function.

iseeyuan replied:
I assume Tup() is a simple helper function. Currently the vector is not used later but not guaranteed. In addition, it's in the export process where I don't think the perf saving (if there is any) with rvalue ref would be noticeable. To keep it simple I'd prefer one API for the helper function.

dreiss commented:
Can you import this to phabricator so we can see the moved code detection?

kimishpatel commented on torch/csrc/jit/mobile/register_mobile_ops.cpp:
```diff
@@ -164,6 +164,40 @@ void format_kernel(const c10::OperatorHandle& op, Stack* stack) {
   push(*stack, ss.str());
 }
 
+void to_dtype_kernal(const c10::OperatorHandle& op, Stack* stack) {
```
Typo, kernal vs kernel.

kimishpatel commented on torch/csrc/jit/mobile/register_mobile_ops.cpp:
```diff
@@ -164,6 +164,40 @@ void format_kernel(const c10::OperatorHandle& op, Stack* stack) {
+       norm_index > static_cast<int64_t>(tuple->elements().size())) {
+     throw std::out_of_range("Tuple list index out of range");
+   }
+   stack->emplace_back(tuple->elements()[norm_index]);
```
Is there a reason to not use `pack` variant here?

kimishpatel reviewed (changes requested):
Left some comments. BTW, for quantized models to work we will also have to add non dynamic versions of linear_run and linear_prepack etc. ops. It is orthogonal to this PR but just wanted to bring it up.

Commit:
Update on "[Lite Interpreter] Enable __setstate__ "


1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)

Commit:
Update on "[Lite Interpreter] Enable __setstate__ "


1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)

kimishpatel reviewed (approved):
LGTM.

Commit:
Update on "[Lite Interpreter] Enable __setstate__ "


1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)

zdevito commented on torch/csrc/jit/mobile/interpreter.cpp:
```diff
@@ -93,6 +93,11 @@ bool InterpreterState::run(Stack& stack) {
         auto userObj = pop(stack).toObject();
+        // Mobile only: since the number of slots is not known, resize the numAttributes
+        // before setSlot.
+        while (userObj->type()->numAttributes() <= inst.X) {
```
The combination of `while`, and `v.type()` is not correct. I think it would be best to just set all member types to `AnyType` rather than give them a wrong type.

iseeyuan replied:
There's exception when I use AnyType. Should I disable `checkNoAny()`?

iseeyuan replied:
I have changed it to NonType, which may be close to "neglecting type in loaded obj" as @zdevito suggested. I feel it's easier than introducing another #ifdef inside setslot(), where some codes may be hidden from unit test.

zdevito commented on torch/csrc/jit/mobile/module.cpp:
```diff
@@ -22,6 +22,15 @@ void CompilationUnit::register_function(std::unique_ptr<Function> fn) {
   methods_.emplace_back(std::move(fn));
 }
 
+Function* CompilationUnit::find_method_by_qn(const c10::QualifiedName& qn) {
```
_by_qn feels redundant and they are Funcitons not methods, so maybe just `find_function`.

zdevito commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -144,69 +226,15 @@ class ScriptModuleSerializer {
-      elements.push_back(Tup({func.qualname().qualifiedName(), table}));
-    }
-    writeArchive("bytecode", Tup(std::move(elements)));
+    // top-level methods
```
This approach will not find all the setstate methods that need to be saved. For instance, if there is a custom class as an attribute of a Module, this will not save it. The correct approach is already implemented for standard serialization: it collects all the classes/functions that have been referenced and makes sure to save them. This code should reuse the result of that analysis to know what __setstate__ methods to produce.

iseeyuan replied:
Good catch! I've changed it the similar way as in `pickler`, to check `ivalue` of the `module` and its `slots`. `checkHasValidSetGetState()` is reused.

zdevito reviewed (commented):
Thanks -- see the inline comment about how to find all the setstate methods, I believe this will miss some.

Commit:
Update on "[Lite Interpreter] Enable __setstate__ "


1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)

Commit:
Update on "[Lite Interpreter] Enable __setstate__ "


1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)

Commit:
Update on "[Lite Interpreter] Enable __setstate__ "


1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)

Commit:
Update on "[Lite Interpreter] Enable __setstate__ "


1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)

Commit:
Update on "[Lite Interpreter] Enable __setstate__ "


1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)

Commit:
Update on "[Lite Interpreter] Enable __setstate__ "


1. Serialize bytecode of __setstate__ and run it when loading the model. 
2. One use case is quantization. To test this use case a few operators are registered temporarily for lite interpreter. The "_" prefix registration will be removed when the operators are all migrated to mobile.

Differential Revision: [D20162898](https://our.internmc.facebook.com/intern/diff/D20162898)
